### PR TITLE
`masked_fill` casts `where` result to `a.dtype`

### LINKED ISSUE #1083: `thunder.jit`ted `Tensor.masked_fill` and `Tensor.masked_fill_` return `torcdh.float32` tensor even when input is `torch.int64`
*Note*: If you have a model or program that is not supported yet but should be, please use the program coverage template.

## 🐛 Bug

As per title.

### To Reproduce

```python
import torch
import thunder


if __name__ == "__main__":
    with torch.device("cuda"):
        a = torch.randint(0, 64, size=(384, 16), dtype=torch.int64)
        b = a.clone().detach()
        mask = torch.randn(size=(384, 16)).to(torch.bool)

    def f(a, mask):
        return a.masked_fill(mask, 0.0)

    expected = f(a, mask)
    jitted = thunder.jit(f)
    actual = jitted(a, mask)

    print(f"# `Tensor.masked_fill`: {expected.dtype = }, {actual.dtype = }")
```

```console
$ python a.py
# `Tensor.masked_fill`: expected.dtype = torch.int64, actual.dtype = torch.float32
```

### Expected behavior

Returns the same dtype as the pytorch result. 

### Environment

- PyTorch Version (e.g., 1.0):
- OS (e.g., Linux):
- How you installed PyTorch (`conda`, `pip`, source):
- Build command you used (if compiling from source):
- Python version:
- CUDA/cuDNN version:
- GPU models and configuration:
- Any other relevant information:

### PR DESCRIPTION
**Pull request by crcrpar**
## What does this PR do?

Fixes #1083.

### CODE REVIEW

File: thunder/torch/__init__.py [hunk 1]
```diff
@@ -2235,8 +2235,8 @@ def _mask_tensor(a, mask, fill_value):
 # NOTE We have chosen not to emulate PyTorch's odd type promotion behavior for this operation
 @torchsymbol(torch.masked_fill, is_method=True)
 def masked_fill(a: TensorLike, /, mask: TensorLike, value: NumberLike | TensorLike) -> TensorLike:
-    result = where(mask, value, a)
-    return result
+    maybe_cast_result = where(mask, value, a)
+    return prims.convert_element_type(maybe_cast_result, a.dtype)
```
Comment (t-vi) [inline thunder/torch/__init__.py:2239]: Wouldn't it be better to cast value as needed to avoid a precision loss in values of a that are not filled?

Comment (crcrpar) [reply thunder/torch/__init__.py:2239]: how could we identify a cast is needed?
With a simple set of dtypes, the current looks fair to me:
```python
# `Tensor.masked_fill`, dtype = torch.int32: expected.dtype = torch.int32, actual.dtype = torch.int32
# Constructed by Delete Last Used (took 0 milliseconds)
import torch
from thunder.executors.torchex import no_autocast

@torch.no_grad()
@no_autocast
def computation(a, mask):
  # a: "cuda:0 i32[384, 16]"
  # mask: "cuda:0 b8[384, 16]"
  [t2] = nvFusion0(a, mask)
    # t0 = prims.convert_element_type(a, dtypes.float32_)  # t0: "cuda:0 f32[384, 16]"
    # t1 = prims.where(mask, 0.0, t0)  # t1: "cuda:0 f32[384, 16]"
    # t2 = prims.convert_element_type(t1, dtypes.int32)  # t2: "cuda:0 i32[384, 16]"
  del a, mask
  return t2
# `Tensor.masked_fill`, dtype = torch.int64: expected.dtype = torch.int64, actual.dtype = torch.int64
# Constructed by Delete Last Used (took 0 milliseconds)
import torch
from thunder.executors.torchex import no_autocast

@torch.no_grad()
@no_autocast
def computation(a, mask):
  # a: "cuda:0 i64[384, 16]"
  # mask: "cuda:0 b8[384, 16]"
  [t2] = nvFusion0(a, mask)
    # t0 = prims.convert_element_type(a, dtypes.float32_)  # t0: "cuda:0 f32[384, 16]"
    # t1 = prims.where(mask, 0.0, t0)  # t1: "cuda:0 f32[384, 16]"
    # t2 = prims.convert_element_type(t1, dtypes.int64)  # t2: "cuda:0 i64[384, 16]"
  del a, mask
  return t2
# `Tensor.masked_fill`, dtype = torch.float64: expected.dtype = torch.float64, actual.dtype = torch.float64
# Constructed by Delete Last Used (took 0 milliseconds)
import torch
from thunder.executors.torchex import no_autocast

@torch.no_grad()
@no_autocast
def computation(a, mask):
  # a: "cuda:0 f64[384, 16]"
  # mask: "cuda:0 b8[384, 16]"
  [t0] = nvFusion0(mask, a)
    # t0 = prims.where(mask, 0.0, a)  # t0: "cuda:0 f64[384, 16]"
  del mask, a
  return t0
# `Tensor.masked_fill`, dtype = torch.bfloat16: expected.dtype = torch.bfloat16, actual.dtype = torch.bfloat16
# Constructed by Delete Last Used (took 0 milliseconds)
import torch
from thunder.executors.torchex import no_autocast

@torch.no_grad()
@no_autocast
def computation(a, mask):
  # a: "cuda:0 bf16[384, 16]"
  # mask: "cuda:0 b8[384, 16]"
  [t0] = nvFusion0(mask, a)
    # t0 = prims.where(mask, 0.0, a)  # t0: "cuda:0 bf16[384, 16]"
  del mask, a
  return t0
```

Comment (t-vi) [reply thunder/torch/__init__.py:2239]:
```
    # t0 = prims.convert_element_type(a, dtypes.float32_)  # t0: "cuda:0 f32[384, 16]"
    # t1 = prims.where(mask, 0.0, t0)  # t1: "cuda:0 f32[384, 16]"
    # t2 = prims.convert_element_type(t1, dtypes.int32)  # t2: "cuda:0 i32[384, 16]"
```
I think this is broken because it will change int numbers that are not exactly representable as fp32 even if when mask is false.

```
a = torch.full((3,), 2**31-1, dtype=torch.int)
a.masked_fill_(torch.arange(3) < 1, 0.0)
```

Comment (crcrpar) [reply thunder/torch/__init__.py:2239]: For clarity, the first `prims.convert_element_type` creating `t0` comes from `where`, not `masked_fill`

Comment (t-vi) [reply thunder/torch/__init__.py:2239]: Yes, but so we need to convert the fill argument to be compatible with the `a.dtype` before entering the where.
